### PR TITLE
Add bullet and torsional friction DOM

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

## Summary

Add Bullet, Torsional friction DOM classes

## Test it
Run the `UNIT_Surface_TEST` and `INTEGRATION_surface_dom` tests

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: include/sdf/Surface.hh [hunk 1]
```diff
@@ -122,6 +122,132 @@ namespace sdf
...
+    public: double Friction2() const;
+
+    /// \brief Set friction coefficient in second friction pyramid direction.
+    /// \param[in] _fricton Friction coefficient
+    public: void SetFriction2(double _friction);
+
+    /// \brief Get the first friction pyramid direction in collision-fixed
+    /// reference
+    /// \return First friction pyramid direction.
+    public: const gz::math::Vector3d &Fdir1() const;
```
Comment (ahcorde) [inline include/sdf/Surface.hh:159]: include `gz/mat/vector3.hh`

File: src/Surface_TEST.cc [hunk 2]
```diff
@@ -322,34 +454,26 @@ TEST(DOMfriction, CopyAssignmentAfterMove)
...
-  EXPECT_DOUBLE_EQ(friction1.ODE()->Slip2(), 4);
-  EXPECT_EQ(friction1.ODE()->Fdir1(),
-            gz::math::Vector3d(1, 2, 3));
+  EXPECT_DOUBLE_EQ(0.3, friction2.Bullet()->Friction());
+  EXPECT_DOUBLE_EQ(0.5, friction2.Bullet()->Friction2());
+  EXPECT_EQ(gz::math::Vector3d(2, 1, 4), friction2.Bullet()->Fdir1());
+  EXPECT_DOUBLE_EQ(1.3, friction2.Bullet()->RollingFriction());
+  EXPECT_DOUBLE_EQ(0.1, friction1.Bullet()->Friction());
+  EXPECT_DOUBLE_EQ(0.2, friction1.Bullet()->Friction2());
+  EXPECT_EQ(gz::math::Vector3d(3, 4, 5), friction1.Bullet()->Fdir1());
```
Comment (ahcorde) [inline src/Surface_TEST.cc:464]: include `gz/math/vector3.hh`

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by azeey** [comment]
@ahcorde mind taking another look

File: src/Surface.cc [hunk 3]
```diff
@@ -63,18 +64,74 @@ class sdf::ODE::Implementation
...
+};
+
 class sdf::Friction::Implementation
 {
-  /// \brief The object storing contact parameters
+  /// \brief The object storing ode parameters
   public: sdf::ODE ode;
 
+  /// \brief The object storing bullet parameters
+  public: std::optional<sdf::Bullet> bullet;
```
Comment (ahcorde) [inline src/Surface.cc:123]: `#include <optional>, <cstdint>`

File: src/Surface.cc [hunk 4]
```diff
@@ -63,18 +64,74 @@ class sdf::ODE::Implementation
...
+
+  /// \brief The object storing torsional parameters
+  public: std::optional<sdf::Torsional> torsional;
```
Comment (ahcorde) [inline src/Surface.cc:126]: include Tosional and butter sdf headers

**Review by ahcorde — APPROVED** [review]
Just two comments, otherwise LGTM

File: src/Surface.cc (hunk 3 above)
Comment (iche033) [reply src/Surface.cc]: added [5378598](https://github.com/gazebosim/sdformat/pull/1351/commits/537859894230dfcfc47e4e7dc9477273b2abf71e).

File: src/Surface.cc (hunk 4 above)
Comment (iche033) [reply src/Surface.cc:129]: the classes are defined in `Surface.hh` so included.

**Review by ahcorde — CHANGES_REQUESTED** [review]
linters 

https://github.com/gazebosim/sdformat/actions/runs/8351884160/job/22861035083?pr=1351

**Comment by iche033** [comment]
I added one more change in [534164b](https://github.com/gazebosim/sdformat/pull/1351/commits/534164b6dd72f00c13c42324edf35c0ada512e0e) that renames the `sdf::Bullet` class to `sdf::BulletFriction` to avoid conflicting with other <bullet> sdf elements in the sdf spec that potentially may have its own DOM classes. 

Side note, there is also the proposal in https://github.com/gazebosim/sdformat/issues/508 to change the way physics engine params are specified, which would affect DOM classes in the future.

**Comment by iche033** [comment]
> linters
>
> https://github.com/gazebosim/sdformat/actions/runs/8351884160/job/22861035083?pr=1351

should be fixed. [4ca98b5](https://github.com/gazebosim/sdformat/pull/1351/commits/4ca98b58ce72c189bfabc782100a54de2a88d799)

**Review by ahcorde — APPROVED** [review]

**Pull request merged**